ministryofjustice/analytics-platform-control-panel-frontend
# Use openid-client for authentication and refreshing tokens

## What

* Remove dependency on `passport-auth0-openidconnect` and add dependency on `openid-client` - this gives us OIDC protocol compliant authentication and userinfo requests, and provides a simple API for refreshing expired tokens.
* Refactor `authentication` middleware to use `openid-client` with Passport so that everything continues to work
* Add expired token handler to `control_panel_api` client, which refreshes the token and retries the request. This should eliminate "Signature has expired" errors. 

## How to review

0. Make sure your `AUTH0_CLIENT_ID` and `AUTH0_CLIENT_SECRET` env vars (and corresponding `OIDC_CLIENT_` vars) are using the `kubectl-oidc` client details in Auth0 dev. (This may require you to update your `~/.kube/config` for RBAC - ask @djsd123 for details - you may also need to set your kubectl context to the dev cluster).
1. Before you checkout this branch, login to the control panel, wait more than 5 mins, reload the home page, and see an error message containing "Signature has expired".
2. Checkout this branch, run and login to the control panel
3. Wait at least 5 mins
4. Reload the home page and see no "Signature has expired" error message.
5. Also see a `info control_panel_api: token expired, refreshing` log message in your console
